packages/grid/x-data-grid/src/hooks/features/headerFiltering/useGridHeaderFiltering.ts

```diff
@@ -0,0 +1,120 @@
+import * as React from 'react';
+import { GridPrivateApiCommunity } from '../../../models/api/gridApiCommunity';
+import { DataGridProcessedProps } from '../../../models/props/DataGridProps';
+import { GridHeaderFilteringState } from '../../../models/gridHeaderFilteringModel';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridStateInitializer } from '../../utils/useGridInitializeState';
+import { useGridLogger } from '../../utils';
+import {
+  gridColumnLookupSelector,
+  gridColumnVisibilityModelSelector,
+  gridColumnFieldsSelector,
+} from '../columns/gridColumnsSelector';
+import {
+  GridHeaderFilteringApi,
+  GridHeaderFilteringPrivateApi,
+} from '../../../models/api/gridHeaderFilteringApi';
+
+export const headerFilteringStateInitializer: GridStateInitializer = (state) => ({
+  ...state,
+  headerFiltering: { editing: null, menuOpen: null },
+});
+
+export const useGridHeaderFiltering = (
+  apiRef: React.MutableRefObject<GridPrivateApiCommunity>,
+  props: Pick<DataGridProcessedProps, 'signature'>,
+) => {
+  const logger = useGridLogger(apiRef, 'useGridHeaderFiltering');
+
+  const setHeaderFilterState = React.useCallback(
+    (headerFilterState: Partial<GridHeaderFilteringState>) => {
+      apiRef.current.setState((state) => {
+        // Safety check to avoid MIT users from using it
+        // This hook should ultimately be moved to the Pro package
+        if (props.signature === 'DataGrid') {
+          return state;
+        }
+        return {
+          ...state,
+          headerFiltering: {
+            editing: headerFilterState.editing ?? null,
+            menuOpen: headerFilterState.menuOpen ?? null,
+          },
+        };
+      });
+      apiRef.current.forceUpdate();
+    },
+    [apiRef, props.signature],
+  );
+
+  const startHeaderFilterEditMode = React.useCallback<
+    GridHeaderFilteringApi['startHeaderFilterEditMode']
+  >(
+    (field) => {
+      logger.debug(`Starting edit mode on header filter for field: ${field}`);
+      apiRef.current.setHeaderFilterState({ editing: field });
+    },
+    [apiRef, logger],
+  );
+
+  const stopHeaderFilterEditMode = React.useCallback<
+    GridHeaderFilteringApi['stopHeaderFilterEditMode']
+  >(() => {
+    logger.debug(`Stopping edit mode on header filter`);
+    apiRef.current.setHeaderFilterState({ editing: null });
+  }, [apiRef, logger]);
+
+  const showHeaderFilterMenu = React.useCallback<GridHeaderFilteringApi['showHeaderFilterMenu']>(
+    (field) => {
+      logger.debug(`Opening header filter menu for field: ${field}`);
+      apiRef.current.setHeaderFilterState({ menuOpen: field });
```

Comment:
I think because we need to check for it in useGridKeyboardNavigation
https://github.com/MBilalShafi/mui-x/blob/5002c7aebb3be0c8bce516a354ac88847588628d/packages/grid/x-data-grid/src/hooks/features/keyboardNavigation/useGridKeyboardNavigation.ts#L317-L318

Reply:
No, I was just wondering if it's indeed necessary. I didn't notice that we need it for keyboard navigation 👍